### ISSUE TITLE
Avoid setting random seed in main tests

### ISSUE THREAD

**Issue by rcurtin** [mlpack/mlpack#1227]
This one is related to #1152 and would be a good issue for someone who is looking to contribute.

One of the keys about the test `mlpack_test` is that its results are reproducible for a single build.  So, the random seed should never be set during the test, instead preferring to leave it to whatever it is set to during compilation.  This means I can run `mlpack_test` multiple times, and if a single random test failed, it will fail again.  This is important for debugging.

However, the recent addition of the main tests in #1152 is actually setting the random seed in many cases to `std::time(NULL)`, which is different for every run.  That breaks the idea set out above... so we should fix it. :)

I would suggest the following steps as a good way to approach this issue:

 * Read through the code to understand how `math::RandomSeed(std::time(NULL))` is being called by some of the main tests in `src/mlpack/tests/*.cpp`.  The call will be in `mlpackMain()` of some of the algorithms being tested.

 * See that `math::RandomSeed()` is actually being called during the tests, perhaps by adding some printed debugging output to the call or setting a breakpoint or something.

 * Think about a way where, if we are in the tests (`mlpack_test` only), we can either avoid calling `math::RandomSeed()` altogether, or make sure that it doesn't actually set the random seed.

 * Implement it!

 * Test that setting the random seed still works for the command-line programs and Python bindings by manually running a program that has the `--seed` option with different seeds, and seeing that the output is different (k-means is one good option).

Regardless of whose PR is merged in the end, I think it is a great learning experience for the codebase for multiple people to work on this, so there is no need to "claim" the issue.  I'm likely to review and merge the first working PR that is submitted.

**Comment by Prabhat-IIT** [comment]
@rcurtin here we have to avoid calling `math::RandomSeed(std::time(NULL))` or both `math::RandomSeed((size_t) CLI::GetParam<int>("seed"))` and `math::RandomSeed(std::time(NULL))`. I'm confused if we avoid calling both i.e. avoid calling `math::RandomSeed()` altogether then how will the randomseed will be set? What should be it's value when we don't pass the `seed` as input Param as well as when we are not allowing the `seed` to be it's default value i.e. 0 ?
If we've to just not allow the calling of `math::RandomSeed()` for `std::time(NULL)` then I think we should decide a fixed value of `seed` whenever `mlpackMain()` is called from any of the `mlpack_tests`. To do this I've two things in mind:

- We can pass a default argument to `mlpackMain()` which will be set to true when we are calling it through any of the `mlpack_test` otherwise which will be false and we can then check its value in `if` condition while calling `RandomSeed()` and if its `true` then we can call `RandomSeed()` for the fixed value of `seed`

-  Plz refer [this](https://stackoverflow.com/questions/353180/how-do-i-find-the-name-of-the-calling-function) Qn. We can find out the calling function name through a static function defined in class `Reporter` which will return a boolean variable indicating if the calling function is one of the `mlpack_test`

**Comment by rcurtin** [comment]
One way would be to disable the functionality of `math::RandomSeed()` when we are calling from the tests.  There are some macros that are set only inside of the tests.

**Comment by Prabhat-IIT** [comment]
Yeah `BINDING_TYPE == BINDING_TYPE_TEST` so we can use this condition to disable the functionality of `math::RandomSeed()`
